### ISSUE TITLE
Create a network diagram

### ISSUE THREAD

**Issue by jeremymanning** [ContextLab/giblet-responses#18]
Use [visualkeras](https://github.com/paulgavrikov/visualkeras/) to create a 3D depiction of the autoencoder network:

```python
import visualkeras

< construct pytorch model >

visualkeras.layered_view(model, legend=True, sizing_mode='logarithmic', show_dimension=True)
```

We may need to play around with the text/options to get the scaling right, avoid overlapping text, ensure clean appearance, etc.

**Comment by jeremymanning** [comment]
note: this should be implemented as part of the toolbox

then call the function and output the figure to papers/figs/source/network.pdf

**Comment by jeremymanning** [comment]
# Implementation Plan: Network Diagram Visualization (Issue #18)

## Overview

Create a publication-quality 3D visualization of the multimodal autoencoder architecture using visualkeras, integrated as a toolbox utility function.

## Requirements Analysis

**From issue #18:**
- Use visualkeras library
- 3D layered view of the network
- Logarithmic sizing for readability (2B parameter model)
- Show dimensions
- Include legend
- Output to `papers/figs/source/network.pdf`
- Part of the toolbox (not standalone script)

**Additional requirements:**
- Clean appearance (no overlapping text)
- Proper scaling
- Color coding by layer type
- Labels for key layers (bottleneck, voxels, modalities)

## Current Architecture Review

**Model structure (from giblet/models/):**
- **MultimodalEncoder**: Video, Audio, Text → Bottleneck (8,000) → Voxels (85,810)
- **MultimodalDecoder**: Voxels/Bottleneck → Video, Audio, Text
- **MultimodalAutoencoder**: Combines encoder + decoder

**Key architectural features to highlight:**
- 3 modality-specific encoders (Video, Audio, Text)
- Bottleneck layer (8,000 dims - middle/smallest)
- fMRI voxel representation (85,810 dims)
- Symmetric decoder paths

## Implementation Plan

### 1. Create Visualization Module

**Location:** `giblet/utils/visualization.py`

**Function:**
```python
def create_network_diagram(
    model,
    output_path='papers/figs/source/network.pdf',
    sizing_mode='logarithmic',
    show_dimensions=True,
    **kwargs
) -> None:
    """
    Create 3D visualization of the autoencoder architecture.
    
    Uses visualkeras to generate publication-quality network diagram.
    """
```

**Features:**
- Load actual MultimodalAutoencoder model
- Configure visualkeras with optimal settings
- Handle PyTorch → visualkeras conversion (if needed)
- Add custom labels/annotations
- Export to PDF (vector format for papers)
- Also export PNG for quick viewing

###2. Test visualkeras with Real Model

**No mocks - use actual autoencoder:**

```python
from giblet.models import create_autoencoder
from giblet.utils.visualization import create_network_diagram

# Create REAL model (2.0B params)
model = create_autoencoder()

# Generate diagram
create_network_diagram(
    model,
    output_path='papers/figs/source/network.pdf'
)
```

**Test requirements:**
- Verify visualkeras handles large model (2B params)
- Check PDF is generated and readable
- Verify all layers visible
- Check dimensions are labeled correctly
- Ensure no overlapping text
- Test different sizing modes if needed

### 3. Optimize Visualization Settings

**Iterative refinement (all with REAL model):**

1. **Test sizing modes:**
   - 'logarithmic' (default)
   - 'linear' (may not work for 2B params)
   - Custom scaling function if needed

2. **Test text placement:**
   - Adjust font sizes
   - Layer spacing
   - Dimension label positions
   - Legend placement

3. **Color customization:**
   - Different colors for encoder vs decoder
   - Highlight bottleneck layer
   - Modality-specific colors (video, audio, text)

4. **Additional annotations:**
   - Layer names (Layer 1-11 from issue #2)
   - Parameter counts
   - Data flow arrows

### 4. Create papers/ Directory Structure

```
papers/
└── figs/
    ├── source/
    │   └── network.pdf     # Vector format (visualkeras output)
    └── compiled/
        └── network.png     # Rasterized for quick viewing
```

### 5. Testing Strategy (NO MOCKS)

**Test file:** `tests/utils/test_visualization.py`

**Tests with REAL model:**

1. **test_create_diagram_basic()**
   - Load real MultimodalAutoencoder
   - Generate diagram to temp file
   - Verify file created
   - Check file size > 0
   - Verify PDF format

2. **test_create_diagram_custom_path()**
   - Test custom output paths
   - Verify directory creation
   - Test with/without file extension

3. **test_diagram_with_different_sizing()**
   - Test sizing_mode options
   - Verify all produce valid outputs
   - Check file sizes reasonable

4. **test_diagram_shows_all_components()**
   - Parse PDF (if possible) or manual verification
   - Check encoder/decoder both present
   - Verify layer count matches architecture

5. **test_visualization_reproducible()**
   - Generate same diagram twice
   - Compare outputs (should be identical)

6. **test_large_model_handling()**
   - Use actual 2B param model
   - Verify doesn't crash or timeout
   - Check memory usage acceptable

**Edge cases:**
- Very deep model
- Model with custom dimensions
- Missing output directory
- Read-only filesystem (should error gracefully)

### 6. Documentation

**Docstrings:**
- Complete function documentation
- Usage examples
- Parameter descriptions
- Return value specification

**User guide:** Add section to README or create `docs/visualization.md`

**Example script:** `examples/create_network_diagram.py`

### 7. Integration

**Update giblet/utils/__init__.py:**
```python
from .visualization import create_network_diagram
```

**Update requirements:**
- Add visualkeras to requirements.txt
- Test installation compatibility

## Dependencies

**New:**
- `visualkeras` - Main visualization library
- `pillow` - Image processing (if not already installed)
- `aggdraw` - For better rendering (optional)

**Existing:**
- PyTorch (already installed)
- matplotlib (already in requirements)

## Deliverables Checklist

- [ ] giblet/utils/visualization.py implemented
- [ ] visualkeras added to requirements
- [ ] papers/figs/source/ directory created
- [ ] network.pdf generated from REAL model
- [ ] network.png for quick viewing
- [ ] tests/utils/test_visualization.py with 6+ tests
- [ ] All tests use REAL MultimodalAutoencoder (no mocks)
- [ ] Example script in examples/
- [ ] Documentation updated
- [ ] Manual verification of diagram quality

## Success Criteria

1. **Functional:**
   - ✅ Function creates diagram from real model without errors
   - ✅ PDF file generated successfully
   - ✅ All tests pass using real model

2. **Quality:**
   - ✅ All layers visible and identifiable
   - ✅ Dimensions shown clearly
   - ✅ No overlapping text
   - ✅ Bottleneck clearly marked
   - ✅ Legend helpful and accurate

3. **Integration:**
   - ✅ Part of giblet.utils module
   - ✅ Can be imported and called easily
   - ✅ Works in fresh conda environment

## Estimated Effort

**Small-Medium:** Straightforward visualization task, main challenge is tuning visualkeras parameters for optimal output with large model.

**Timeline:** 1-2 hours (including iterative refinement)

## Notes

- visualkeras is designed for Keras, but should work with PyTorch models
- May need adapter if direct PyTorch support is limited
- Could fall back to custom matplotlib visualization if visualkeras problematic
- Focus on publication quality - this goes in papers

**Comment by jeremymanning** [comment]
✅ Issue #18 complete!

Network diagram visualization implemented and tested.

**Deliverables:**
- Custom PyTorch visualization (giblet/utils/visualization.py)
- Publication-quality diagrams (papers/figs/source/network.pdf, network.png)
- 25 tests, all passing with REAL 2.0B param model
- CLI tool (examples/create_network_diagram.py)

**Quality:**
- 3D layered view with depth effect
- Logarithmic sizing (handles massive parameter ranges)
- Color-coded modalities
- All 52 layers visible, no overlapping text
- Ready for publication

**Commit:** b545639

**Comment by jeremymanning** [comment]
## ✅ PlotNeuralNet Implementation Complete

All identified issues fixed with professional PlotNeuralNet solution.

### Issues Resolved

1. ✅ **Directory corrected:** papers/ → paper/
2. ✅ **Professional quality:** LaTeX/TikZ output (readable text, clean layout)
3. ✅ **Horizontal orientation:** Left-to-right flow
4. ✅ **Parallel branches clearly shown:** 2A/B/C and 10A/B/C visible

### Deliverables

**Generated diagram:**
- `paper/figs/source/network.pdf` (42KB, publication-ready)
- `paper/figs/source/network.tex` (LaTeX source, editable)

**Code:**
- `giblet/utils/plotneuralnet.py` (PlotNeuralNet integration)
- `examples/generate_network_diagram.py` (one-command generation)

**Usage:**
```bash
python examples/generate_network_diagram.py
```

### Architecture Shown

All 11 layers with parallel processing:
- Layer 1: 3 inputs (video/audio/text)
- Layer 2A/B/C: Parallel encoders (video conv/audio conv/text linear)
- Layer 3: Feature pooling
- Layer 4: Feature conv + ReLU
- Layer 5: Linear to voxels (85,810)
- Layer 6: Bottleneck (8,000 - middle/smallest)
- Layers 7-11: Symmetric decoder with parallel outputs

### Quality Verification

✓ PDF successfully compiled (42KB)  
✓ Horizontal left-to-right flow  
✓ Parallel branches visible  
✓ Professional LaTeX/TikZ styling  
✓ Editable .tex source provided  

**Commit:** c44f2fb

---

**Status:** Ready for publication use. Please review `paper/figs/source/network.pdf`.
